espnet2/uasr/espnet_model.py

```diff
@@ -0,0 +1,476 @@
+            text.shape,
+            text_lengths.shape,
+        )
+        batch_size = speech.shape[0]
+
+        # for data-parallel
+        text = text[:, : text_lengths.max()]
+
+        # 1. Feats encode (Extract feats + Apply segmenter)
+        feats, padding_mask = self.encode(speech, speech_lengths)
+
+        # 2. Generate fake samples
+        (
+            generated_sample,
+            real_sample,
+            x_inter,
+            generated_sample_padding_mask,
+        ) = self.generator(feats, text, padding_mask)
+
+        # 3. Reprocess segments
+        if self.use_segmenter:
+            (
+                generated_sample,
+                generated_sample_padding_mask,
+            ) = self.segmenter.logit_segment(
+                generated_sample, generated_sample_padding_mask
+            )
+
+        # for phone_diversity_loss
+        generated_sample_logits = generated_sample
+
+        if not self.skip_softmax:
+            if self.training and self.use_gumbel:
+                generated_sample = F.gumbel_softmax(
+                    generated_sample_logits.float(),
+                    tau=self.curr_temp,
+                    hard=self.use_hard_gumbel,
+                ).type_as(generated_sample_logits)
+            else:
+                generated_sample = generated_sample_logits.softmax(-1)
+
+        # for validation
+        vocab_seen = None
+        if do_validation:
+            batch_num_errors = 0
+            batched_hyp_ids = generated_sample.argmax(-1)
+            batched_hyp_ids[generated_sample_padding_mask] = self.pad
+
+            # for kenlm ppl metric
+            batch_lm_log_prob = 0
+            batch_num_hyp_tokens = 0
+            vocab_seen = torch.zeros(self.vocab_size - 4, dtype=torch.bool)
+
+            for hyp_ids, ref_ids in zip(batched_hyp_ids, text):
+                # remove <pad> and <unk>
+                hyp_ids = hyp_ids[hyp_ids >= 4]
+                # remove duplicate tokens
+                hyp_ids = hyp_ids.unique_consecutive()
+                # remove silence
+                hyp_ids_nosil = hyp_ids[hyp_ids != self.sil_id]
+                hyp_ids_nosil_list = hyp_ids_nosil.tolist()
+
+                if self.kenlm:
+                    hyp_token_list = self.token_id_converter.ids2tokens(
+                        integers=hyp_ids
+                    )
+                    hyp_tokens = " ".join(hyp_token_list)
+                    lm_log_prob = self.kenlm.score(hyp_tokens)
+                    batch_lm_log_prob += lm_log_prob
+                    batch_num_hyp_tokens += len(hyp_token_list)
+
+                    hyp_tokens_index = hyp_ids[hyp_ids >= 4]
+                    vocab_seen[hyp_tokens_index - 4] = True
+
+                ref_ids = ref_ids[ref_ids != self.pad]
+                ref_ids_list = ref_ids.tolist()
+                num_errors = editdistance.eval(hyp_ids_nosil_list, ref_ids_list)
+                batch_num_errors += num_errors
```

Comment:
I feel like it would be better to keep the dev set to have text for consistency reasons (even some fake texts). I suggest adding an option for showing the PER since it is meaningless for fake texts.